src/types/permissions.ts

```diff
@@ -42,33 +50,151 @@ export enum PermissionType {
+  readonly type = PermissionType.FungibleAsset;
+
+  readonly asset: AccountAddress;
+
+  readonly balance: string;
+
+  constructor(asset: AccountAddress, balance: string) {
+    super();
+    this.asset = asset;
+    this.balance = balance;
+  }
+
+  static from({ asset, balance }: { asset: AccountAddress; balance: string }): FungibleAssetPermission {
+    return new FungibleAssetPermission(asset, balance);
+  }
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    this.asset.serialize(serializer);
+    serializer.serializeStr(this.balance);
+  }
+
+  static deserialize(deserializer: Deserializer): FungibleAssetPermission {
+    deserializer.deserializeStr() as PermissionType.FungibleAsset; // type, is this needed?;
+    const asset = AccountAddress.deserialize(deserializer);
+    const balance = deserializer.deserializeStr();
+    return new FungibleAssetPermission(asset, balance);
+  }
 }
 
-export interface GasPermission {
-  type: PermissionType.Gas;
-  amount: number;
+export class GasPermission extends Serializable {
+  readonly type = PermissionType.Gas;
+
+  readonly amount: number;
+
+  constructor(amount: number) {
+    super();
+    this.amount = amount;
+  }
+
+  static from = ({ amount }: { amount: number }): GasPermission => new GasPermission(amount);
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    serializer.serializeU16(this.amount);
+  }
+
+  static deserialize(deserializer: Deserializer): GasPermission {
+    deserializer.deserializeStr() as PermissionType.Gas; // type, is this needed?;
+    const amount = deserializer.deserializeU16();
+    return new GasPermission(amount);
+  }
 }
 
-export interface NFTPermission {
-  type: PermissionType.NFT;
-  assetAddress: string;
-  capabilities: Record<NFTCapability, boolean>;
+export class NFTPermission extends Serializable {
+  readonly type = PermissionType.NFT;
+
+  readonly assetAddress: AccountAddress;
+
+  readonly capabilities: Record<NFTCapability, boolean>; // (string[], bool[])
+
+  constructor(assetAddress: AccountAddress, capabilities: Record<NFTCapability, boolean>) {
+    super();
+    this.assetAddress = assetAddress;
+    this.capabilities = capabilities;
+  }
+
+  static from = ({
+    assetAddress,
+    capabilities,
+  }: {
+    assetAddress: AccountAddress;
+    capabilities: Record<NFTCapability, boolean>;
+  }): NFTPermission => new NFTPermission(assetAddress, capabilities);
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    this.assetAddress.serialize(serializer);
+
+    const [capabilityKeys, capabilityValues] = Object.entries(this.capabilities).reduce(
+      ([keys, values], [key, value]) => [keys.concat(key), values.concat(value)],
+      [[] as string[], [] as boolean[]],
+    );
+    serializer.serializeStr(JSON.stringify(capabilityKeys));
+    serializer.serializeStr(JSON.stringify(capabilityValues));
```

Comment:
Serializing the capabilities is a bit messy. The base type is:

```
capabilities: {
mutate: true,
transfer: false
}
``` 
where each property is optional. Let me know of a better way.